B2-40 support 100 Continue

emnoor-reef opened the pull request:
The `b2sdk/urllib3/_connection.py` file is mostly copied from [`botocore`](https://github.com/boto/botocore/blob/754b699bbf34261eae47c9dece3b11d7b58eb03c/botocore/awsrequest.py). I have updated at some places to support urllib3>=2 (botocore pins urllib3<1.27).

mjurbanski-reef commented on test/integration/test_raw_api.py:
```diff
@@ -550,6 +551,71 @@ def raw_api_test_helper(raw_api, should_cleanup_old_buckets):
+                server_side_encryption=sse_b2_aes,
+            )
+
+        # Check if data was sent
+        assert file_contents not in sent_data
```
I'm somewhat confused;
wouldn't it be probable that data wasn't sent just because invalidauthtoken error?

emnoor-reef replied:
I didn't get your question.

By default, `http.client` doesn't care if `Expect: 100-continue` header is set and sends the entire body with the request. When reading for response it just ignores `100 Continue` responses and keeps waiting for non-1XX response.

We are intercepting the request/response cycles when `Except: 100-continue` is present. So this test just verifies that we do not send the body initially, and we do not send the body at all if we do not receive `100 Continue` response.

mjurbanski-reef replied:
shouldn't it always fallback to sending data after a timeout?

mjurbanski-reef commented on test/integration/test_raw_api.py:
```diff
@@ -550,6 +551,71 @@ def raw_api_test_helper(raw_api, should_cleanup_old_buckets):
+            {'color': 'blue'},
+            data,
+            server_side_encryption=sse_b2_aes,
+        )
```
please add `assert file_contents in sent_data` otherwise it very well may be that you are not patching the right function

mjurbanski-reef commented on b2sdk/urllib3/NOTICE:
```diff
@@ -0,0 +1,8 @@
+Urllib3
```
this module name says that it vendors urllib3 when in reality it vendors botocore and not urllib3 at all

botocore patches both urllib3 and requests

suggested name `_botocore`.
* botocore since it is botocore code
* `_` since we don't want anybody EVER think they should do `import b2sdk._botocore` (apiver already forbids this, but `_` prefix is more well known convention than apiver

mjurbanski-reef commented on b2sdk/urllib3/NOTICE:
```diff
@@ -0,0 +1,8 @@
+
+Copyright 2023 Backblaze Inc.
+Changes made to the original source:
+The following classes have been overriden for support HTTP 100-continue:
+urllib3.connection.HTTPConnection, urllib3.connection.VerifiedHTTPSConnection,
+urllib3.connectionpool.HTTPConnectionPool, urllib3.connectionpool.HTTPSConnectionPool
```
these are not our (Backblaze) changes as you are suggesting thru the header above!

IMO we should just say we are vendoring part of botocore that is relavant to supporting 100 Continue and list our changes to it below it.

mjurbanski-reef commented on b2sdk/b2http.py:
```diff
@@ -22,8 +22,8 @@
 
 import requests
-from requests.adapters import HTTPAdapter
 
+from b2sdk.urllib3._connection import HTTPAdapterWithContinue
```
I got scared a little that we are switching from requests to raw urllib, but apparently `b2sdk.urllib3` module name is just wrong here

mjurbanski-reef commented on test/integration/test_raw_api.py:
```diff
@@ -550,6 +551,71 @@ def raw_api_test_helper(raw_api, should_cleanup_old_buckets):
             is_file_lock_enabled=False,
         )
 
+    # b2_100_continue
```
I don't like the way this test is done (multiple test cases in a single test, and mocking in Integration tests...), but I guess you inherited it ;p
Lets just make sure it working correctly.

mjurbanski-reef commented on b2sdk/urllib3/_connection.py:
```diff
@@ -0,0 +1,238 @@
+            # This is our custom behavior.  If the Expect header was
+            # set, it will trigger this custom behavior.
+            logger.debug("Waiting for 100 Continue response.")
+            # Wait for 1 second for the server to send a response.
+            if urllib3.util.wait_for_read(self.sock, 2):
```
lol, this is a good case showing why putting an obvious is decremental
it says 1 second, but wait is actually 2 ;p

as to why I'm commenting on this: we may need this to be user-configurable
https://github.com/reef-technologies/b2-sdk-python/pull/252/files#diff-ac825af986ae146afcaa98597de5a9c07399afb45b45edbba3d2b7ab68db124eR26

Please for now just expose it so it is controllable for HTTPAdapterWithContinue user and use `10` as default since that value was chosen with pycurl for some reason.

mjurbanski-reef commented on test/integration/test_raw_api.py:
```diff
@@ -550,6 +551,71 @@ def raw_api_test_helper(raw_api, should_cleanup_old_buckets):
+
+    def set_100_header(fn):
+        def wrapper(url, headers, *args, **kwargs):
+            headers.update({
+                'Expect': '100-continue',
+            })
```
wait, what, we are not setting this header by default during upload etc? 
I wonder what is the purpose of adding this in if we are not using it

please consult this with @pawelpolewicz  if this should be enabled by default or turned off completely for now

I'm leaving out the "optional" option since IMO having it is for "greater good" (infrastructure wise ;) ) and having it always on may only negatively affect very small file uploads which are not the main use case of b2 AFAIK.

mjurbanski-reef reviewed (changes requested):
I like reusing the opensource components here :muscle: 
this is why I left most of the comments on License-related files - nitpicks mostly, but lets make sure we are playing by the rules correctly :)

emnoor-reef commented on b2sdk/urllib3/_connection.py:
```diff
@@ -0,0 +1,238 @@
+# Code taken from:
+# https://github.com/boto/botocore/blob/754b699bbf34261eae47c9dece3b11d7b58eb03c/botocore/awsrequest.py
+# The code has been modified to work with urllib3>=2.0
```
yes, it works with both urllib3 1.x and 2.x. I have added a dependency to urllib3 for b2sdk.

mjurbanski-reef commented on b2sdk/_botocore/awsrequest.py:
```diff
@@ -82,6 +83,11 @@ def request(self, method, url, body=None, headers=None, *args, **kwargs):
         self._response_received = False
         if headers.get('Expect', b'') in [b'100-continue', '100-continue']:
             self._expect_header_set = True
+            timeout = headers.pop('X-Expect-100-Continue-Timeout-Seconds', self._continue_timeout)
```
nitpick: drop `-Seconds` part of the name; AFAIK there are already timeout settigns http headers, none of which adds `-seconds` (Cache-Control, Keep Alive etc)

mjurbanski-reef commented on b2sdk/_botocore/awsrequest.py:
```diff
@@ -82,6 +83,11 @@ def request(self, method, url, body=None, headers=None, *args, **kwargs):
+            timeout = headers.pop('X-Expect-100-Continue-Timeout-Seconds', self._continue_timeout)
+            try:
+                self._continue_timeout = float(timeout)
+            except (ValueError, TypeError):
+                pass
```
personally, I would prefer it explode; name collision seems unlikely for it to be breaking change

mjurbanski-reef commented on b2sdk/api_config.py:
```diff
@@ -26,7 +26,9 @@ def __init__(
-        decode_content: bool = False
+        decode_content: bool = False,
+        expect_100_continue: bool = True,
+        expect_100_continue_timeout_seconds: float = 10.0,
```
please rename to `expect_100_timeout` across the whole code base and make sure you added it to docstrings to methods which already have params in docstrings

Lets not do "expect 100 continue timeout" in one place and "expect 100 timeout"  in another. Lets have one and stick with it.

mjurbanski-reef commented on b2sdk/file_version.py:
```diff
@@ -351,6 +351,7 @@ def _get_upload_headers(self) -> bytes:
             file_retention=self.file_retention,
             legal_hold=self.legal_hold,
             cache_control=self.cache_control,
+            expect_100_continue=False,
```
wasn't it supposed to be ON by default with option of disabling?

emnoor-reef replied:
Isn't `file_version` a separate request than an upload request?

mjurbanski-reef replied:
sorry my bad:)

also this isn't used in any request but  in `has_large_header` from what I see

mjurbanski-reef reviewed (changes requested):
solution looks good now, but I'm pretty sure there is still something shady going on in the test

also make sure default is set properly &  add it to changelog please

mjurbanski-reef commented on NOTES:
```diff
@@ -0,0 +1,5 @@
+response status line = b'HTTP/1.1 401 \r\n'
```
:wastebasket:

mjurbanski-reef commented on b2sdk/_botocore/awsrequest.py:
```diff
@@ -0,0 +1,229 @@
+            # This is our custom behavior.  If the Expect header was
+            # set, it will trigger this custom behavior.
+            logger.debug("Waiting for 100 Continue response.")
+            # Wait for 10 seconds for the server to send a response.
```
remove this comment :) it is not accurate depending on configuration

emnoor-reef commented:
I agree the test is a bit shady.

In the last commit I moved the test into a separate file. I added two more tests for the the configurations. Though I am still not happy with the tests.

Also, the fixture could be broken up into reusable parts. Currently it borrows all of it's code form `test_raw_api`. But I think that could be refactored with `test_raw_api` when that is done.

mjurbanski-reef commented on test/integration/test_raw_expect_100.py:
```diff
@@ -0,0 +1,151 @@
+    }
+
+
+@pytest.fixture(autouse=True)
+def rest_sent_data(expect_100_setup):
+    expect_100_setup["sent_data"].clear()
```
it kinda seems to me like this test will work even if this was deleted

mjurbanski-reef commented on test/integration/test_raw_expect_100.py:
```diff
@@ -0,0 +1,151 @@
+    file_sha1 = hex_sha1_of_stream(io.BytesIO(file_contents), file_length)
+
+    return {
+        "raw_api": raw_api,
+        "upload_url": upload_url,
+        "upload_auth_token": upload_auth_token,
+        "sse_b2_aes": sse_b2_aes,
+        "sent_data": sent_data,
+        "file_name": file_name,
+        "file_contents": file_contents,
+        "file_length": file_length,
+        "file_sha1": file_sha1,
```
seems like you crammed bunch of fixtures into single one
this doesn't encourage reusability

mjurbanski-reef commented on b2sdk/raw_api.py:
```diff
@@ -949,6 +959,8 @@ def upload_file(
         :param legal_hold: legal hold setting for the file
         :param custom_upload_timestamp: custom upload timestamp for the file
         :param cache_control: an optional cache control setting. Syntax based on the section 14.9 of RFC 2616. Example string value: 'public, max-age=86400, s-maxage=3600, no-transform'.
+        :param expect_100_continue: whether to use 'Expect: 100-continue' header
+        :param expect_100_timeout: timeout of 100 response when expect_100_continue is True
```
sorry for another change here, but now that I look at it - wouldn't `expect_100_timeout` being set to 0 be basically the same think and we could skip adding `expect_100_continue` param altogether?

mjurbanski-reef commented on test/integration/test_raw_expect_100.py:
```diff
@@ -0,0 +1,151 @@
+    assert expect_100_setup["file_contents"] not in expect_100_setup["sent_data"]
+
+
+def test_expect_100_timeout(expect_100_setup):
```
and where is the test that we actually don't send data before 100 Continue is received from the server, but after we get it, we do send data? That is the most important test; The "ignoring 100 Continue" is what we did so far without modification to urrlib&requests

mjurbanski-reef commented on test/integration/test_raw_expect_100.py:
```diff
@@ -0,0 +1,151 @@
+        account_id, int(time.time()), random.randint(1000, 9999)
+    )
+
+    bucket_dict = raw_api.create_bucket(
+        api_url,
+        account_auth_token,
+        account_id,
+        bucket_name,
+        'allPublic',
+        is_file_lock_enabled=True,
+    )
```
please reuse the single bucket for all tests here - there is not reason not to, and it will reduce the overall API calls i.e. make tests faster

mjurbanski-reef commented on test/integration/test_raw_expect_100.py:
```diff
@@ -0,0 +1,93 @@
+        expect_100_timeout=0,
+    )
+    assert file_contents in http_sent_data
+
```
test proving expect 100 continue works with b2 server is still missing

mjurbanski-reef commented on test/unit/botocore/test_awsrequest.py:
```diff
@@ -0,0 +1,348 @@
+        response = conn.getresponse()
+        self.assertEqual(response.status, 200)
+
+    def test_tunnel_readline_none_bugfix(self):
+        # Tests whether ``_tunnel`` function is able to work around the
+        # py26 bug of avoiding infinite while loop if nothing is returned.
+        conn = self.create_tunneled_connection(
+            url='s3.amazonaws.com',
+            port=443,
+            response=b'HTTP/1.1 200 OK\r\n',
+        )
+        conn._tunnel()
+        # Ensure proper amount of readline calls were made.
+        self.assertEqual(self.mock_response.fp.readline.call_count, 2)
```
lets throw out test that are not useful to us so we don't waste time on maintenence

mjurbanski-reef commented on test/integration/test_raw_expect_100.py:
```diff
@@ -0,0 +1,137 @@
+        )
+    assert file_contents in http_sent_data
+    assert wait_mock.call_count == 1
+    args, _ = wait_mock.call_args
```
whats the purpose of this?

emnoor-reef replied:
forgot to remove that line :}

mjurbanski-reef commented on test/unit/botocore/test_awsrequest.py:
```diff
@@ -0,0 +1,348 @@
+            self.assertIn(b'body', s.sent_data)
+
+
+if __name__ == "__main__":
+    unittest.main()
```
also throw this out

mjurbanski-reef reviewed (approved):
https://github.com/reef-technologies/b2-sdk-python/pull/300/files#r1297715473 would be nice to include, but not exactly crucial
